Indexing multiple files per Solr document (n:1)

Hi, 
I have been trying to index multiple files per Solr document (n:1) following the [docs](https://dbmdz.github.io/solr-ocrhighlighting/indexing/#multiple-files-per-solr-document-n1) and got this error. If I index them **one file per Solr document (1:1)**, there is **no problem**.

This is the request body with the concatenated files:
```json
{
    "id": "es-scbg_bblg_18950101",
    "ocr_text": "/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0008.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0003.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0002.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0001.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0005.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0004.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0006.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0007.xml"
}
```

And the following response:
```json
{
    "responseHeader": {
        "status": 400,
        "QTime": 80
    },
    "error": {
        "metadata": ["error-class", "org.apache.solr.common.SolrException", "root-error-class", "com.ctc.wstx.exc.WstxUnexpectedCharException"],
        "msg": "Exception writing document id es-scbg_bblg_18950101 to the index; possible analysis error.",
        "code": 400
    }
}
```

Schema file:
```xml
<schema name="coordinateHighlight" version="1.0">
  <types>
    <fieldType name="string" class="solr.StrField" sortMissingLast="true" omitNorms="true"/>
    <fieldtype name="text_ocr" class="solr.TextField" storeOffsetsWithPositions="true" termVectors="true">
      <analyzer type="index">
        <charFilter class="de.digitalcollections.solrocr.lucene.filters.ExternalUtf8ContentFilterFactory" />
        <charFilter class="de.digitalcollections.solrocr.lucene.filters.OcrCharFilterFactory" />
        <tokenizer class="solr.StandardTokenizerFactory"/>
        <filter class="solr.LowerCaseFilterFactory"/>
        <filter class="solr.StopFilterFactory"/>
        <filter class="solr.PorterStemFilterFactory"/>
      </analyzer>
      <analyzer type="query">
        <tokenizer class="solr.StandardTokenizerFactory"/>
        <filter class="solr.LowerCaseFilterFactory"/>
        <filter class="solr.StopFilterFactory"/>
        <filter class="solr.PorterStemFilterFactory"/>
      </analyzer>
    </fieldtype>
  </types>

  <fields>
    <field name="id" type="string" multiValued="false" indexed="true" stored="true"/>
    <field name="ocr_text" type="text_ocr" multiValued="false" indexed="true" stored="true"/>
  </fields>
  <uniqueKey>id</uniqueKey>
</schema>
```

Solr log:
```bash
solr  | 2021-05-31 08:00:46.285 INFO  (qtp722951168-23) [   x:ocr] o.a.s.u.p.LogUpdateProcessorFactory [ocr]  webapp=/solr path=/update/json/docs params={commit=true}{} 0 72
solr  | 2021-05-31 08:00:46.288 ERROR (qtp722951168-23) [   x:ocr] o.a.s.h.RequestHandlerBase org.apache.solr.common.SolrException: Exception writing document id es-scbg_bblg_18950101 to the index; possible analysis error.
solr  |         at org.apache.solr.update.DirectUpdateHandler2.addDoc(DirectUpdateHandler2.java:256)
solr  |         at org.apache.solr.update.processor.RunUpdateProcessorFactory$RunUpdateProcessor.processAdd(RunUpdateProcessorFactory.java:73)
solr  |         at org.apache.solr.update.processor.UpdateRequestProcessor.processAdd(UpdateRequestProcessor.java:55)
solr  |         at org.apache.solr.update.processor.DistributedUpdateProcessor.versionAdd(DistributedUpdateProcessor.java:288)
solr  |         at org.apache.solr.update.processor.DistributedUpdateProcessor.processAdd(DistributedUpdateProcessor.java:222)
solr  |         at org.apache.solr.update.processor.LogUpdateProcessorFactory$LogUpdateProcessor.processAdd(LogUpdateProcessorFactory.java:106)
solr  |         at org.apache.solr.handler.loader.JsonLoader$SingleThreadedJsonLoader$1.handle(JsonLoader.java:241)
solr  |         at org.apache.solr.common.util.JsonRecordReader$Node.handleObjectStart(JsonRecordReader.java:391)
solr  |         at org.apache.solr.common.util.JsonRecordReader$Node.parse(JsonRecordReader.java:292)
solr  |         at org.apache.solr.common.util.JsonRecordReader$Node.access$200(JsonRecordReader.java:156)
solr  |         at org.apache.solr.common.util.JsonRecordReader.streamRecords(JsonRecordReader.java:141)
solr  |         at org.apache.solr.handler.loader.JsonLoader$SingleThreadedJsonLoader.handleSplitMode(JsonLoader.java:221)
solr  |         at org.apache.solr.handler.loader.JsonLoader$SingleThreadedJsonLoader.processUpdate(JsonLoader.java:136)
solr  |         at org.apache.solr.handler.loader.JsonLoader$SingleThreadedJsonLoader.load(JsonLoader.java:121)
solr  |         at org.apache.solr.handler.loader.JsonLoader.load(JsonLoader.java:84)
solr  |         at org.apache.solr.handler.UpdateRequestHandler$1.load(UpdateRequestHandler.java:97)
solr  |         at org.apache.solr.handler.ContentStreamHandlerBase.handleRequestBody(ContentStreamHandlerBase.java:82)
solr  |         at org.apache.solr.handler.RequestHandlerBase.handleRequest(RequestHandlerBase.java:216)
solr  |         at org.apache.solr.core.SolrCore.execute(SolrCore.java:2646)
solr  |         at org.apache.solr.servlet.HttpSolrCall.execute(HttpSolrCall.java:794)
solr  |         at org.apache.solr.servlet.HttpSolrCall.call(HttpSolrCall.java:567)
solr  |         at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:427)
solr  |         at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:357)
solr  |         at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:201)
solr  |         at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)
solr  |         at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:548)
solr  |         at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
solr  |         at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:602)
solr  |         at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
solr  |         at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
solr  |         at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1612)
solr  |         at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
solr  |         at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1434)
solr  |         at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
solr  |         at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
solr  |         at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1582)
solr  |         at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
solr  |         at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1349)
solr  |         at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
solr  |         at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
solr  |         at org.eclipse.jetty.server.handler.InetAccessHandler.handle(InetAccessHandler.java:177)
solr  |         at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
solr  |         at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
solr  |         at org.eclipse.jetty.rewrite.handler.RewriteHandler.handle(RewriteHandler.java:322)
solr  |         at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
solr  |         at org.eclipse.jetty.server.Server.handle(Server.java:516)
solr  |         at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
solr  |         at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
solr  |         at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
solr  |         at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
solr  |         at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
solr  |         at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
solr  |         at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
solr  |         at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
solr  |         at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
solr  |         at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
solr  |         at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
solr  |         at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
solr  |         at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:773)
solr  |         at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:905)
solr  |         at java.base/java.lang.Thread.run(Unknown Source)
solr  | Caused by: java.lang.RuntimeException: Failed to parse the OCR markup, make sure your files are well-formed and your regions start/end on complete tags! (Source was: /data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0008.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0003.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0002.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0001.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0005.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0004.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0006.xml+/data/es-scbg_bblg_18950101/alto/es-scbg_bblg_18950101_0007.xml)
solr  |         at de.digitalcollections.solrocr.formats.OcrParser.next(OcrParser.java:144)
solr  |         at de.digitalcollections.solrocr.lucene.filters.OcrCharFilter.readNextWord(OcrCharFilter.java:29)
solr  |         at de.digitalcollections.solrocr.lucene.filters.OcrCharFilter.read(OcrCharFilter.java:125)
solr  |         at org.apache.lucene.analysis.standard.StandardTokenizerImpl.zzRefill(StandardTokenizerImpl.java:675)
solr  |         at org.apache.lucene.analysis.standard.StandardTokenizerImpl.getNextToken(StandardTokenizerImpl.java:898)
solr  |         at org.apache.lucene.analysis.standard.StandardTokenizer.incrementToken(StandardTokenizer.java:148)
solr  |         at org.apache.lucene.analysis.LowerCaseFilter.incrementToken(LowerCaseFilter.java:41)
solr  |         at org.apache.lucene.analysis.FilteringTokenFilter.incrementToken(FilteringTokenFilter.java:49)
solr  |         at org.apache.lucene.analysis.en.PorterStemFilter.incrementToken(PorterStemFilter.java:67)
solr  |         at org.apache.lucene.index.DefaultIndexingChain$PerField.invert(DefaultIndexingChain.java:921)
solr  |         at org.apache.lucene.index.DefaultIndexingChain.processField(DefaultIndexingChain.java:524)
solr  |         at org.apache.lucene.index.DefaultIndexingChain.processDocument(DefaultIndexingChain.java:488)
solr  |         at org.apache.lucene.index.DocumentsWriterPerThread.updateDocuments(DocumentsWriterPerThread.java:208)
solr  |         at org.apache.lucene.index.DocumentsWriter.updateDocuments(DocumentsWriter.java:415)
solr  |         at org.apache.lucene.index.IndexWriter.updateDocuments(IndexWriter.java:1471)
solr  |         at org.apache.lucene.index.IndexWriter.updateDocuments(IndexWriter.java:1464)
solr  |         at org.apache.solr.update.DirectUpdateHandler2.updateDocOrDocValues(DirectUpdateHandler2.java:967)
solr  |         at org.apache.solr.update.DirectUpdateHandler2.doNormalUpdate(DirectUpdateHandler2.java:342)
solr  |         at org.apache.solr.update.DirectUpdateHandler2.addDoc0(DirectUpdateHandler2.java:294)
solr  |         at org.apache.solr.update.DirectUpdateHandler2.addDoc(DirectUpdateHandler2.java:241)
solr  |         ... 60 more
solr  | Caused by: com.ctc.wstx.exc.WstxUnexpectedCharException: Unexpected character '' (code 65279 / 0xfeff) in epilog; expected '<'
solr  |  at [row,col {unknown-source}]: [1389,8]
solr  |         at com.ctc.wstx.sr.StreamScanner.throwUnexpectedChar(StreamScanner.java:647)
solr  |         at com.ctc.wstx.sr.BasicStreamReader.nextFromProlog(BasicStreamReader.java:2054)
solr  |         at com.ctc.wstx.sr.BasicStreamReader.closeContentTree(BasicStreamReader.java:2911)
solr  |         at com.ctc.wstx.sr.BasicStreamReader.nextFromTree(BasicStreamReader.java:2654)
solr  |         at com.ctc.wstx.sr.BasicStreamReader.next(BasicStreamReader.java:1073)
solr  |         at de.digitalcollections.solrocr.formats.alto.AltoParser.seekToNextWord(AltoParser.java:228)
solr  |         at de.digitalcollections.solrocr.formats.alto.AltoParser.readNext(AltoParser.java:153)
solr  |         at de.digitalcollections.solrocr.formats.OcrParser.next(OcrParser.java:140)
solr  |         ... 79 more
```

Thanks!

--- a/pom.xml
+++ b/pom.xml
@@ -54,7 +54,6 @@
     <version.byteseek>2.0.3</version.byteseek>
     <version.commons-text>1.9</version.commons-text>
     <version.junit>5.8.2</version.junit>
-    <version.log4j>2.11.1</version.log4j>
     <version.mockito>4.1.0</version.mockito>
     <version.slf4j>1.7.32</version.slf4j>
     <version.solr>8.9.0</version.solr>

--- a/src/main/java/de/digitalcollections/solrocr/lucene/filters/ExternalUtf8ContentFilterFactory.java
+++ b/src/main/java/de/digitalcollections/solrocr/lucene/filters/ExternalUtf8ContentFilterFactory.java
@@ -33,8 +33,8 @@ import org.apache.solr.common.SolrException.ErrorCode;
 public class ExternalUtf8ContentFilterFactory extends CharFilterFactory {
   public ExternalUtf8ContentFilterFactory(Map<String, String> args) {
     super(args);
-    // TODO: Read whitelisted base directories from config
-    // TODO: Read whitelisted file name patterns from config
+    // TODO: Read allowed base directories from config
+    // TODO: Read allowed filename patterns from config
     // TODO: Warn of security implications if neither is defined
   }
 
@@ -88,15 +88,14 @@ public class ExternalUtf8ContentFilterFactory extends CharFilterFactory {
     } catch (IOException e) {
       throw new RuntimeException(
           String.format(
-              Locale.US, "Error while reading external content from pointer '%s': %s", ptrStr, e));
+              Locale.US, "Error while reading external content from pointer '%s': %s", ptrStr, e),
+          e);
     }
   }
 
   private void validateSource(SourcePointer.FileSource src) {
-    // TODO: Check if sourcePath is located under one of the whitelisted base directories, abort
-    // otherwise
-    // TODO: Check if sourcePath's filename matches one of the whitelisted file name patterns, abort
-    // otherwise
+    // TODO: Check if sourcePath is located under one of the allowed base directories, else abort
+    // TODO: Check if sourcePath's filename matches one of the allowed filename patterns, else abort
     File f = src.path.toFile();
     if (!f.exists() || !f.canRead()) {
       throw new SolrException(
@@ -139,13 +138,24 @@ public class ExternalUtf8ContentFilterFactory extends CharFilterFactory {
     int charOffset = 0;
     ByteBuffer buf = ByteBuffer.allocateDirect(1024 * 1024 /* 1 MiB */);
     // TODO: Use a queue for the file sources so we don't have to read until the end of the last
-    // file every time
+    //       file every time
     // TODO: Think about building the UTF8 -> UTF16 offset map right here if the mapping part should
-    // become a
-    //       bottle neck
+    //       become a bottle neck
     for (SourcePointer.FileSource src : ptr.sources) {
       try (FileChannel fChan = FileChannel.open(src.path, StandardOpenOption.READ)) {
         final int fSize = (int) fChan.size();
+
+        int bomOffset = 0;
+        if (!src.isAscii) {
+          // Check for BOM, we need to skip it as to not break mult-file parsing
+          ByteBuffer bomBuf = ByteBuffer.allocate(3);
+          fChan.read(bomBuf, 0);
+          bomBuf.flip();
+          if (bomBuf.equals(ByteBuffer.wrap(new byte[] {(byte) 0xEF, (byte) 0xBB, (byte) 0xBF}))) {
+            bomOffset = 3;
+          }
+          fChan.position(0);
+        }
 
         // Byte offset of the current file from the beginning of the first file
         final int baseOffset = byteOffset;
@@ -158,6 +168,10 @@ public class ExternalUtf8ContentFilterFactory extends CharFilterFactory {
             region.start += baseOffset;
             region.end = Math.min(region.end + baseOffset, fSize + baseOffset);
             continue;
+          }
+          if (region.start == 0) {
+            // Skip the BOM at the start of a file, if present
+            region.start += bomOffset;
           }
           // Make region offsets relative to the beginning of the first file
           region.start += baseOffset;

--- a/src/test/resources/data/chronicling_hocr/seq-1.html
+++ b/src/test/resources/data/chronicling_hocr/seq-1.html
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
 <!DOCTYPE html
   PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
 <html xmlns="http://www.w3.org/1999/xhtml">

--- a/src/test/resources/data/chronicling_hocr/seq-2.html
+++ b/src/test/resources/data/chronicling_hocr/seq-2.html
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
 <!DOCTYPE html
   PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
 <html xmlns="http://www.w3.org/1999/xhtml">

--- a/src/test/resources/data/chronicling_hocr/seq-3.html
+++ b/src/test/resources/data/chronicling_hocr/seq-3.html
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
 <!DOCTYPE html
   PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
 <html xmlns="http://www.w3.org/1999/xhtml">

--- a/src/test/resources/data/chronicling_hocr/seq-4.html
+++ b/src/test/resources/data/chronicling_hocr/seq-4.html
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
 <!DOCTYPE html
   PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
 <html xmlns="http://www.w3.org/1999/xhtml">

